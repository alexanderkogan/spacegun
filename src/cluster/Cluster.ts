import { Image } from "@/images/ImageProvider";

export interface ClusterProvider {
    clusters: string[]
    pods(cluster: string): Promise<Pod[]>
    deployments(cluster: string): Promise<Deployment[]>
    scalers(cluster: string): Promise<Scaler[]>
}

export interface Pod {
    name: string
    image: Image
    restarts: number
}

export interface Deployment {
    name: string
    image: Image
}

export interface Scaler {
    name: string
    replicas: Replicas
}

export interface Replicas {
    current: number
    minimum: number
    maximum: number
}
