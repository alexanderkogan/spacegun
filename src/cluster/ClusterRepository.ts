import { Pod } from './model/Pod'
import { Image } from './model/Image'
import { Deployment } from './model/Deployment'
import { Scaler } from './model/Scaler'
import { ServerGroup } from './model/ServerGroup'
import { ClusterSnapshot } from './model/ClusterSnapshot'

export interface ClusterRepository {
    clusters: string[]

    namespaces(context: string): Promise<string[]>
    pods(group: ServerGroup): Promise<Pod[]>
    deployments(group: ServerGroup): Promise<Deployment[]>
    updateDeployment(group: ServerGroup, deployment: Deployment, targetImage: Image): Promise<Deployment>
    restartDeployment(group: ServerGroup, deployment: Deployment): Promise<Deployment>
    scalers(group: ServerGroup): Promise<Scaler[]>

    takeSnapshot(group: ServerGroup): Promise<ClusterSnapshot>
    applySnapshot(group: ServerGroup, snapshot: ClusterSnapshot, ignoreImage: boolean): Promise<void>
}
