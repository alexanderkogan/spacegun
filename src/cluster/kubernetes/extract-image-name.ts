export function extractImageName(url: string): string {
    const endOfHost = url.lastIndexOf('/')
    const imageWithoutHost = url.substring(endOfHost + 1, url.length)
    const imageWithoutHash = imageWithoutHost.split('@')[0]
    const imageAndTag = imageWithoutHash.split(':')
    return imageAndTag[0]
}
