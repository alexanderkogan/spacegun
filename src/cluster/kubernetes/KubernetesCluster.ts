import { ClusterProvider, Pod, Deployment, Scaler } from "../Cluster"
const cloneDeep = require("lodash.clonedeep")

import {
    KubeConfig,
    Core_v1Api, V1PodList,
    Apps_v1beta2Api, V1beta2DeploymentList,
    Autoscaling_v1Api, V1HorizontalPodAutoscalerList
} from '@kubernetes/client-node'

interface Api {
    setDefaultAuthentication(config: KubeConfig): void
}

export class KubernetesClusterProvider implements ClusterProvider {

    private configs: Map<string, KubeConfig>

    public constructor(configFile: string) {
        const config = new KubeConfig()
        config.loadFromFile(configFile)

        this.configs = new Map()
        for (const contexts of config.getContexts()) {
            const clusterConfig = cloneDeep(config)
            clusterConfig.setCurrentContext(contexts.name)
            this.configs.set(contexts.name, clusterConfig)
        }
    }

    get clusters(): string[] {
        return Array.from(this.configs.keys())
    }

    async pods(cluster: string): Promise<Pod[]> {
        const api = this.build(cluster, (server: string) => new Core_v1Api(server))
        const result: V1PodList = await api.listNamespacedPod("default").get("body")
        return result.items.map(item => {
            const image = item.spec.containers[0].image
            const tag = image.split(":")[1]
            return {
                name: item.metadata.name,
                image: { image, tag },
                restarts: item.status.containerStatuses[0].restartCount
            }
        })
    }

    async deployments(cluster: string): Promise<Deployment[]> {
        const api = this.build(cluster, (server: string) => new Apps_v1beta2Api(server))
        const result: V1beta2DeploymentList = await api.listNamespacedDeployment("default").get("body")
        return result.items.map(item => {
            const image = item.spec.template.spec.containers[0].image
            const tag = image.split(":")[1]
            return {
                name: item.metadata.name,
                image: { image, tag }
            }
        })
    }

    async scalers(cluster: string): Promise<Scaler[]> {
        const api = this.build(cluster, (server: string) => new Autoscaling_v1Api(server))
        const result: V1HorizontalPodAutoscalerList = await api.listNamespacedHorizontalPodAutoscaler("default").get("body")
        return result.items.map(item => ({
            name: item.metadata.name,
            replicas: {
                current: item.status.currentReplicas,
                minimum: item.spec.minReplicas,
                maximum: item.spec.maxReplicas
            }
        }))
    }

    private getConfig(cluster: string): KubeConfig {
        const config: KubeConfig | undefined = this.configs.get(cluster)
        if (config === undefined) {
            throw new Error(`Config for cluster ${cluster} could not be found`)
        }
        return config
    }

    private getServer(config: KubeConfig): string {
        return config.getCurrentCluster().server
    }

    private build<T extends Api>(cluster: string, apiProvider: (server: string) => T): T {
        const config: KubeConfig = this.getConfig(cluster)
        const api: T = apiProvider(this.getServer(config))
        api.setDefaultAuthentication(config)
        return api
    }
}
