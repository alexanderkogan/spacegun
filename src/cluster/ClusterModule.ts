import { ClusterRepository } from "./ClusterRepository"
import { Pod } from "./model/Pod"
import { Deployment } from "./model/Deployment"
import { Scaler } from "./model/Scaler"
import { Image } from "./model/Image"
import { ServerGroup } from "./model/ServerGroup"
import { ClusterSnapshot } from "./model/ClusterSnapshot"

import { RequestInput } from "../dispatcher/model/RequestInput"
import { Request } from "../dispatcher/model/Request"
import { Component } from "../dispatcher/component"
import { Layers } from "../dispatcher/model/Layers"
import { Methods } from "../dispatcher/model/Methods"

let repo: ClusterRepository | undefined = undefined
export function init(clusterRepository: ClusterRepository) {
    repo = clusterRepository
}

const serverGroupInput = (input: ServerGroup | undefined) => RequestInput.of(
    ["cluster", input!.cluster],
    ["namespace", input!.namespace]
)
const serverGroupMapper = (p: RequestInput) => ({
    cluster: p.params!["cluster"],
    namespace: p.params!["namespace"]
} as ServerGroup)

export const clusters: Request<void, string[]> = {
    module: "cluster",
    procedure: "clusters"
}

export const namespaces: Request<{ cluster: string }, string[]> = {
    module: "cluster",
    procedure: "namespaces",
    input: (input: { cluster: string } | undefined) => RequestInput.of(["cluster", input!.cluster]),
    mapper: (input: RequestInput) => ({ cluster: input.params!["cluster"] as string })
}

export const updateDeployment: Request<UpdateDeploymentParameters, Deployment> = {
    module: "cluster",
    procedure: "updateDeployment",
    input: (input: UpdateDeploymentParameters | undefined) => RequestInput.ofData(
        { deployment: input!.deployment, image: input!.image },
        ["cluster", input!.group.cluster],
        ["namespace", input!.group.namespace]
    ),
    mapper: (p: RequestInput): UpdateDeploymentParameters => ({
        group: {
            cluster: p.params!["cluster"],
            namespace: p.params!["namespace"]
        } as ServerGroup,
        deployment: p.data.deployment as Deployment,
        image: p.data.image as Image
    })
}

export const restartDeployment: Request<RestartDeploymentParameters, Deployment> = {
    module: "cluster",
    procedure: "restartDeployment",
    input: (input: RestartDeploymentParameters | undefined) => RequestInput.ofData(
        { deployment: input!.deployment },
        ["cluster", input!.group.cluster],
        ["namespace", input!.group.namespace]
    ),
    mapper: (p: RequestInput): RestartDeploymentParameters => ({
        group: {
            cluster: p.params!["cluster"],
            namespace: p.params!["namespace"]
        } as ServerGroup,
        deployment: p.data.deployment as Deployment
    })
}

export const pods: Request<ServerGroup, Pod[]> = {
    module: "cluster",
    procedure: "pods",
    input: serverGroupInput,
    mapper: serverGroupMapper
}

export const deployments: Request<ServerGroup, Deployment[]> = {
    module: "cluster",
    procedure: "deployments",
    input: serverGroupInput,
    mapper: serverGroupMapper
}

export const scalers: Request<ServerGroup, Scaler[]> = {
    module: "cluster",
    procedure: "scalers",
    input: serverGroupInput,
    mapper: serverGroupMapper
}

export const takeSnapshot: Request<ServerGroup, ClusterSnapshot> = {
    module: "cluster",
    procedure: "takeSnapshot",
    input: serverGroupInput,
    mapper: serverGroupMapper
}

export const applySnapshot: Request<ApplySnapshotParameters, void> = {
    module: "cluster",
    procedure: "applySnapshot",
    input: (input: ApplySnapshotParameters | undefined) => RequestInput.ofData(
        { snapshot: input!.snapshot },
        ["cluster", input!.group.cluster],
        ["namespace", input!.group.namespace],
        ["ignoreImage", input!.ignoreImage]
    ),
    mapper: (p: RequestInput) => ({
        snapshot: p.data.snapshot as ClusterSnapshot,
        group: {
            cluster: p.params!["cluster"],
            namespace: p.params!["namespace"]
        } as ServerGroup,
        ignoreImage: p.params!["ignoreImage"] === undefined ? undefined : p.params!["ignoreImage"] === "true"
    } as ApplySnapshotParameters)
}

export interface UpdateDeploymentParameters {
    group: ServerGroup, deployment: Deployment, image: Image
}

export interface RestartDeploymentParameters {
    group: ServerGroup, deployment: Deployment
}

export interface ApplySnapshotParameters {
    group: ServerGroup, snapshot: ClusterSnapshot, ignoreImage?: boolean
}

export class Module {

    @Component({
        moduleName: clusters.module,
        layer: Layers.Server
    })
    [clusters.procedure](): Promise<string[]> {
        return Promise.resolve(repo!.clusters)
    }

    @Component({
        moduleName: namespaces.module,
        layer: Layers.Server,
        mapper: namespaces.mapper
    })
    [namespaces.procedure](params: { cluster: string }): Promise<string[]> {
        return repo!.namespaces(params.cluster)
    }

    @Component({
        moduleName: pods.module,
        layer: Layers.Server,
        mapper: pods.mapper
    })
    [pods.procedure](group: ServerGroup): Promise<Pod[]> {
        return repo!.pods(group)
    }

    @Component({
        moduleName: updateDeployment.module,
        layer: Layers.Server,
        mapper: updateDeployment.mapper,
        method: Methods.Put
    })
    [updateDeployment.procedure](input: UpdateDeploymentParameters): Promise<Deployment> {
        return repo!.updateDeployment(input.group, input.deployment, input.image)
    }

    @Component({
        moduleName: restartDeployment.module,
        layer: Layers.Server,
        mapper: restartDeployment.mapper,
        method: Methods.Put
    })
    [restartDeployment.procedure](input: RestartDeploymentParameters): Promise<Deployment> {
        return repo!.restartDeployment(input.group, input.deployment)
    }

    @Component({
        moduleName: deployments.module,
        layer: Layers.Server,
        mapper: deployments.mapper
    })
    [deployments.procedure](group: ServerGroup): Promise<Deployment[]> {
        return repo!.deployments(group)
    }

    @Component({
        moduleName: scalers.module,
        layer: Layers.Server,
        mapper: scalers.mapper
    })
    [scalers.procedure](group: ServerGroup): Promise<Scaler[]> {
        return repo!.scalers(group)
    }

    @Component({
        moduleName: takeSnapshot.module,
        layer: Layers.Server,
        mapper: takeSnapshot.mapper
    })
    [takeSnapshot.procedure](group: ServerGroup): Promise<ClusterSnapshot> {
        return repo!.takeSnapshot(group)
    }

    @Component({
        moduleName: applySnapshot.module,
        layer: Layers.Server,
        mapper: applySnapshot.mapper,
        method: Methods.Put
    })
    [applySnapshot.procedure](params: ApplySnapshotParameters): Promise<void> {
        return repo!.applySnapshot(params.group, params.snapshot, params.ignoreImage === undefined ? true : params.ignoreImage)
    }
}
