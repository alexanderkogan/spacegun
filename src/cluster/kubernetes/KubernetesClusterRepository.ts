import { Pod } from "@/cluster/model/Pod"
import { Image } from "@/cluster/model/Image"
import { Deployment } from "@/cluster/model/Deployment"
import { Scaler } from "@/cluster/model/Scaler"
import { ClusterRepository } from "@/cluster/ClusterRepository"
import { Cache } from "@/Cache"

const cloneDeep = require("lodash.clonedeep")

import {
    KubeConfig,
    Core_v1Api, V1PodList,
    Apps_v1beta2Api, V1beta2DeploymentList,
    Autoscaling_v1Api, V1HorizontalPodAutoscalerList, V1beta2Deployment, V1Container, V1NamespaceList, V1PodStatus
} from '@kubernetes/client-node'
import { ServerGroup } from "@/cluster/model/ServerGroup"
import { ClusterSnapshot } from "@/cluster/model/ClusterSnapshot";

interface Api {
    setDefaultAuthentication(config: KubeConfig): void
}

class UpdateDeploymentApi extends Apps_v1beta2Api {
    defaultHeaders = { "content-type": "application/merge-patch+json" }

    constructor(server: string) {
        super(server)
    }
}

export class KubernetesClusterRepository implements ClusterRepository {

    private namespacesCache: Cache<string, string[]> = new Cache(60)

    public static fromConfig(configFile: string, namespaces?: string[]): KubernetesClusterRepository {
        const config = new KubeConfig()
        config.loadFromFile(configFile)
        const configs = new Map()
        for (const contexts of config.getContexts()) {
            const clusterConfig = cloneDeep(config)
            clusterConfig.setCurrentContext(contexts.name)
            configs.set(contexts.name, clusterConfig)
        }
        return new KubernetesClusterRepository(configs, namespaces)
    }

    public constructor(
        public readonly configs: Map<string, KubeConfig>,
        private readonly allowedNamespaces: string[] | undefined
    ) { }

    get clusters(): string[] {
        return Array.from(this.configs.keys())
    }

    async namespaces(context: string): Promise<string[]> {
        return this.namespacesCache.calculate(context, async () => {
            const api = this.build(context, (server: string) => new Core_v1Api(server))
            const result: V1NamespaceList = await api.listNamespace().get("body")
            return result.items
                .map(namespace => namespace.metadata.name)
                .filter(namespace => this.isNamespaceAllowed(namespace))
        })
    }

    async pods(group: ServerGroup): Promise<Pod[]> {
        const api = this.build(group.cluster, (server: string) => new Core_v1Api(server))
        const namespace = this.getNamespace(group)
        const result: V1PodList = await api.listNamespacedPod(namespace).get("body")
        return result.items.map(item => {
            const image = this.createImage(item.spec.containers)
            let restarts
            if (item.status.containerStatuses != undefined && item.status.containerStatuses.length >= 1) {
                restarts = item.status.containerStatuses[0].restartCount
            }
            const ready = this.isReady(item.status)
            return {
                name: item.metadata.name,
                image, restarts, ready
            }
        })
    }

    async deployments(cluster: ServerGroup): Promise<Deployment[]> {
        const api = this.build(cluster.cluster, (server: string) => new Apps_v1beta2Api(server))
        const namespace = this.getNamespace(cluster)
        const result: V1beta2DeploymentList = await api.listNamespacedDeployment(namespace).get("body")
        return result.items.map(item => {
            const image = this.createImage(item.spec.template.spec.containers)
            return {
                name: item.metadata.name,
                image
            }
        })
    }

    async scalers(group: ServerGroup): Promise<Scaler[]> {
        const api = this.build(group.cluster, (server: string) => new Autoscaling_v1Api(server))
        const namespace = this.getNamespace(group)
        const result: V1HorizontalPodAutoscalerList = await api.listNamespacedHorizontalPodAutoscaler(namespace).get("body")
        return result.items.map(item => ({
            name: item.metadata.name,
            replicas: {
                current: item.status.currentReplicas,
                minimum: item.spec.minReplicas,
                maximum: item.spec.maxReplicas
            }
        }))
    }

    async updateDeployment(group: ServerGroup, deployment: Deployment, targetImage: Image): Promise<Deployment> {
        const api = this.build(group.cluster, (server: string) => new UpdateDeploymentApi(server))
        const namespace = this.getNamespace(group)
        const patch = {
            apiVersion: "apps/v1beta2",
            kind: "Deployment",
            spec: {
                template: {
                    spec: {
                        containers: [{
                            name: deployment.name,
                            image: targetImage.url,
                        }]
                    }
                }
            }
        }
        const result: V1beta2Deployment = await api.patchNamespacedDeployment(deployment.name, namespace, patch).get("body")
        return {
            name: result.metadata.name,
            image: this.createImage(result.spec.template.spec.containers)
        }
    }

    async takeSnapshot(group: ServerGroup): Promise<ClusterSnapshot> {
        const api = this.build(group.cluster, (server: string) => new Apps_v1beta2Api(server))
        const namespace = this.getNamespace(group)
        const result: V1beta2DeploymentList = await api.listNamespacedDeployment(namespace).get("body")
        return {
            deployments: result.items.map(d => ({
                name: d.metadata.name,
                data: d
            }))
        }
    }

    async applySnapshot(group: ServerGroup, snapshot: ClusterSnapshot): Promise<void> {
        const api = this.build(group.cluster, (server: string) => new Apps_v1beta2Api(server))
        const namespace = this.getNamespace(group)
        for (const deployment of snapshot.deployments) {
            await api.replaceNamespacedDeployment(
                deployment.name,
                namespace,
                deployment.data as V1beta2Deployment
            )
        }
    }

    private getNamespace(group: ServerGroup): string {
        return group.namespace || "default"
    }

    private isNamespaceAllowed(namespace: string): boolean {
        return this.allowedNamespaces === undefined
            || this.allowedNamespaces.find(n => n === namespace) !== undefined
    }

    private isReady(status: V1PodStatus): boolean {
        const readyCondition = status.conditions && status.conditions.find(c => c.type === 'Ready')
        return readyCondition !== undefined && readyCondition.status === 'True'
    }

    private createImage(containers: Array<V1Container> | undefined): Image | undefined {
        if (containers !== undefined && containers.length >= 1) {
            const url = containers[0].image
            const imageAndTag = url.split(":")
            const imageParts = imageAndTag[0].split("/")
            const tag = imageAndTag[1]
            const name = imageParts[imageParts.length - 1]
            return { url, name, tag }
        }
        return undefined
    }

    private getConfig(cluster: string): KubeConfig {
        const config: KubeConfig | undefined = this.configs.get(cluster)
        if (config === undefined) {
            throw new Error(`Config for cluster ${cluster} could not be found`)
        }
        return config
    }

    private getServer(config: KubeConfig): string {
        return config.getCurrentCluster().server
    }

    private build<T extends Api>(cluster: string, apiProvider: (server: string) => T): T {
        const config: KubeConfig = this.getConfig(cluster)
        const api: T = apiProvider(this.getServer(config))
        api.setDefaultAuthentication(config)
        return api
    }
}
