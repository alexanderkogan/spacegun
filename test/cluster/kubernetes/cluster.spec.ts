import { KubernetesClusterProvider } from "../../../src/cluster/kubernetes/KubernetesCluster"
import { Pod, Deployment, Scaler } from "../../../src/cluster/Cluster"

const image1 = { name: "image1", tag: "tag", url: "repo/image1:tag" }
const image2 = { name: "image2", tag: "tag", url: "repo/image2:tag" }

describe("KubernetesClusterProvider", () => {
    const cluster = new KubernetesClusterProvider('./test/config/kube')

    describe("clusters", () => {
        it("returns the names of the clusters", () => {
            expect(cluster.clusters).toEqual(["dev", "pre", "live"])
        })
    })

    describe("pods", () => {
        it("returns pods", async () => {
            const pods: Pod[] = await cluster.pods(cluster.clusters[0])
            expect(pods).toEqual([
                { image: image1, name: "pod1", restarts: 0, ready: true },
                { image: image2, name: "pod2", restarts: 1, ready: false },
            ])
        })
    })

    describe("deployements", () => {
        it("returns deployements", async () => {
            const deployements: Deployment[] = await cluster.deployments(cluster.clusters[0])
            expect(deployements).toEqual([
                { image: image1, name: "deployement1" },
                { image: image2, name: "deployement2" }
            ])
        })

        it("updates deployments", async () => {
            const deployement: Deployment = await cluster.updateDeployment(
                cluster.clusters[0], { image: image1, name: "deployement1" }, image2
            )
            expect(deployement).toEqual(
                { image: { name: "updatedImage", tag: "tag", url: "repo/updatedImage:tag" }, name: "updatedDeployment" }
            )
        })
    })

    describe("scalers", () => {
        it("returns horizontal auto scalers", async () => {
            const scalers: Scaler[] = await cluster.scalers(cluster.clusters[0])
            expect(scalers).toEqual([
                { name: "pod1", replicas: { current: 0, maximum: 2, minimum: 1 } },
                { name: "pod2", replicas: { current: 1, maximum: 3, minimum: 2 } }
            ])
        })
    })
})
