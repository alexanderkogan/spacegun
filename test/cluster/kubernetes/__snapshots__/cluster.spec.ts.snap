// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KubernetesClusterProvider appliesSnapshots for batch jobs calls endpoints on snapshot change 1`] = `
Array [
  "batch1",
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "batch1",
    },
    "spec": Object {
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "metadata": Object {
              "annotations": Object {},
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "image": "repo/image1:tag@some:digest",
                },
              ],
            },
          },
        },
      },
      "schedule": "some schedule",
    },
  },
]
`;

exports[`KubernetesClusterProvider appliesSnapshots for batch jobs creates batch jobs if batch job is not known yet 1`] = `
Array [
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "somesillyjob",
    },
    "spec": Object {
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "metadata": Object {
              "annotations": Object {},
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "image": "somenewsillyimage",
                },
              ],
            },
          },
        },
      },
      "schedule": "someSchedule",
    },
  },
]
`;

exports[`KubernetesClusterProvider batches restarts batches 1`] = `
Array [
  "batch1",
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "batch1",
    },
    "spec": Object {
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "metadata": Object {
              "annotations": Object {
                "spacegun.batch": "1520899200000",
              },
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "image": "repo/image1:tag@some:digest",
                },
              ],
            },
          },
        },
      },
      "schedule": "someSchedule",
    },
  },
]
`;

exports[`KubernetesClusterProvider batches updates batches 1`] = `
Array [
  "batch1",
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "batch1",
    },
    "spec": Object {
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "metadata": Object {
              "annotations": Object {
                "spacegun.batch": "1520899200000",
              },
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "image": "repo/image2:tag@some:digest",
                },
              ],
            },
          },
        },
      },
      "schedule": "someSchedule",
    },
  },
]
`;

exports[`KubernetesClusterProvider deployments restarts deployments 1`] = `
Array [
  "deployment1",
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "deployment1",
    },
    "spec": Object {
      "template": Object {
        "metadata": Object {
          "annotations": Object {
            "spacegun.deployment": "1520899200000",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "repo/image1:tag@some:digest",
            },
          ],
        },
      },
    },
  },
]
`;

exports[`KubernetesClusterProvider deployments updates deployments 1`] = `
Array [
  "deployment1",
  "default",
  Object {
    "metadata": Object {
      "annotations": Object {},
      "name": "deployment1",
    },
    "spec": Object {
      "template": Object {
        "metadata": Object {
          "annotations": Object {
            "spacegun.deployment": "1520899200000",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "image": "repo/image2:tag@some:digest",
            },
          ],
        },
      },
    },
  },
]
`;
